python/lookout/sdk/grpc/interceptors/utils.py

```diff
@@ -0,0 +1,130 @@
+from typing import Union
+
+import grpc
+
+ServerInterceptor = Union[
+    'UnaryServerInterceptor',
+    'StreamServerInterceptork'
+]
+
+
+class ServerInterceptorWrapper(grpc.ServerInterceptor):
+    """gRPC server interceptor wrapper
+
+    Unfortunately at time of writing (`grpcio==1.18.0`) gRPC doesn't provide
+    an api to create a server interceptor for each type of invocation:
+        - unary-unary,
+        - unary-stream,
+        - stream-unary,
+        - stream-stream.
+
+    But it provides a single interface for all methods:
+        https://github.com/grpc/grpc/blob/f6cba4b2fba2fd529d68146a4725b56e806cc6f2/src/python/grpcio/grpc/__init__.py#L1275
+
+    This class is a sort of hack that permits the user to have a clean api to
+    create server interceptors for each gRPC method type. Currently the
+    following classes are provided (similarly to the Go SDK):
+        - `lookout.sdk.grpc.interceptors.utils.UnaryServerInterceptor`,
+        - `lookout.sdk.grpc.interceptors.utils.ServerServerInterceptor`.
+
+    The first one intercepts both unary-unary and stream-unary invocations (so
+    the invocations where the server response is unary).
+    The second one intercepts both unary-stream and stream-stream invocations
+    (so the invocations where the server response is a stream).
+
+    More fine-grainded distinction could be easily added.
+
+    This wrapper only needs the server it is attached to. Everytime a request
+    is intercepted through the `intercept_service` method exposed by the
+    `grpc.ServerInterceptor` interface, the type of the RPC method handler
+    is checked against the type of wrapped interceptor. If the types match,
+    then the interceptor is called, otherwise it is skipped.
+
+    """
+    def __init__(self, wrapped: ServerInterceptor):
+        """Wraps the provided server interceptor"""
+        self._wrapped = wrapped
+        self._server = None
+
+    def bind(self, server: grpc.Server):
+        """Binds the provided server"""
+        self._server = server
+
+    def intercept_service(self, continuation, handler_call_details):
+        """Intercepts incoming RPCs before handing them over to a handler
+
+        See `grpc.ServerInterceptor.intercept_service`.
+
+        """
+        method = handler_call_details.method
+        rpc_method_handler = self._server._state.generic_handlers[0]\
```

Comment:
do we need a check for  `_server is None` here?

Reply:
Yup you're right. In that case I'd raise an error such as `RuntimeError("ServerIntercepor is not bind to a server")`.

Reply:
would be nice to add a comment about `0`. Right now it looks like magic.

Reply:
done